addon/corpus.js

```diff
@@ -1,106 +1,231 @@
-async function freezeAllPages() {
-    // Grey out Freeze button:
-    document.getElementById('freeze').disabled = true;
+let gUrls;  // Array of [filename, url] to freeze.
+let gUrlIndex;  // Pointer to current url in gUrls.
+let gFreezeOptions; // Freeze-dry options from UI.
+let gTimeout; // Load+Freeze timeout from UI.
+
+async function freezeAllPages(event) {
+    event.preventDefault();
+
+    // Check form validity.
+    if (!(
+        document.getElementById('wait').validity.valid &&
+        document.getElementById('timeout').validity.valid
+    )) {
+        return;
+    }
+
+    initialiseFromForm();
+    emptyElement(document.getElementById('status'));
 
-    // Clear error field:
-    const errorField = document.getElementById('errors');
-    while (errorField.firstChild) {
-        errorField.removeChild(errorField.firstChild);
+    // We need at least one url.
+    if (!gUrls.length) {
+        let li = document.createElement('li');
+        li.classList.add('error');
+        li.appendChild(document.createTextNode('no pages to download'));
+        document.getElementById('status').appendChild(li);
+        return;
     }
 
-    // Make freezing window, and set its size. The blank page acts as a
+    document.getElementById('freeze').disabled = true;
+    let windowId = 0;
+
+    // Listen for tab events before we start creating tabs; this avoids race
+    // conditions between creating tabs and waiting for loading to complete.
+    browser.tabs.onUpdated.addListener((tabId, changeInfo, tab) => {
```

Comment:
Is there a descriptive name we could give this listener function? It's pretty long to read to get a gist.